### PR TITLE
Protect against common firstRender issues

### PR DESCRIPTION
**Pull request by trueadm**
Release notes: none

firstRender can fail when `Function.prototype.bind` is passed around because the firstRender removes the class component entirely, along with any concept of state etc. To prevent outputting the incorrect value, we now check for bind and if we're given something that is completely abstract as a child from a render, we bail out in that case too.

### CODE REVIEW

**Comment by gaearon** [comment]
If I explicitly optimize the middle component, then it fails with expected invariant (I don't know how to assert on this in test though):

<details>

```js
var React = require('react');
this['React'] = React;

// FB www polyfill
if (!this.babelHelpers) {
  this.babelHelpers = {
    inherits(subClass, superClass) {
      Object.assign(subClass, superClass);
      subClass.prototype = Object.create(superClass && superClass.prototype);
      subClass.prototype.constructor = subClass;
      subClass.__superConstructor__ = superClass;
      return superClass;
    },
  };
}

var _React$Component = babelHelpers.inherits(
  Middle,
  React.Component
);
_superProto = _React$Component && _React$Component.prototype;

function Middle(props) {
  _superProto.constructor.call(this, props);
  this.$Middle_renderItem = function() {
    return this.props.item;
  }.bind(this);
}
Middle.prototype.render = function() {
  var children = this.props.children;
  return children({
    renderItem: this.$Middle_renderItem,
  });
};

function App(props) {
  return <Middle {...props} />;
}

App.getTrials = function(renderer, Root) {
  renderer.update(
    <Root item={<span>Hi</span>}>
      {obj => <h1>{obj.renderItem()}</h1>}
    </Root>
  );
  return [['render with bound child function', renderer.toJSON()]];
};

if (this.__optimizeReactComponentTree) {
  __optimizeReactComponentTree(Middle, {
    firstRenderOnly: true
  });
}

module.exports = App;
```

</details>

----

I tried to optimize the parent component so that it falls back. But it still fails with this test case—this time, with the actual bug:

<details>

```js
var React = require('react');
this['React'] = React;

// FB www polyfill
if (!this.babelHelpers) {
  this.babelHelpers = {
    inherits(subClass, superClass) {
      Object.assign(subClass, superClass);
      subClass.prototype = Object.create(superClass && superClass.prototype);
      subClass.prototype.constructor = subClass;
      subClass.__superConstructor__ = superClass;
      return superClass;
    },
  };
}

var _React$Component = babelHelpers.inherits(
  Middle,
  React.Component
);
_superProto = _React$Component && _React$Component.prototype;

function Middle(props) {
  _superProto.constructor.call(this, props);
  this.$Middle_renderItem = function() {
    return this.props.item;
  }.bind(this);
}
Middle.prototype.render = function() {
  var children = this.props.children;
  return children({
    renderItem: this.$Middle_renderItem,
  });
};

function App(props) {
  return <Middle {...props} />;
}

App.getTrials = function(renderer, Root) {
  renderer.update(
    <Root item={<span>Hi</span>}>
      {obj => <h1>{obj.renderItem()}</h1>}
    </Root>
  );
  return [['render with bound child function', renderer.toJSON()]];
};

if (this.__optimizeReactComponentTree) {
  __optimizeReactComponentTree(App, {
    firstRenderOnly: true
  });
}

module.exports = App;
```

</details>

File: src/react/reconcilation.js [hunk 1]
```diff
@@ -1040,7 +1045,13 @@ export class Reconciler {
   }
 
   _findReactComponentTrees(value: Value, evaluatedNode: ReactEvaluatedNode): void {
-    if (value instanceof AbstractValue) {
+    if (value instanceof BoundFunctionValue) {
```
Comment (sebmarkbage) [inline src/react/reconcilation.js]: It's not just bound functions. It's all functions that are not ECMAScriptSourceFunctionValue, right?

`value instanceof FunctionValue && !(value instanceof ECMAScriptSourceFunctionValue)`

Comment (sebmarkbage) [reply src/react/reconcilation.js]: This pattern match should probably be exhaustive. A whitelist instead of blacklist. What values are valid here?

**Comment by sebmarkbage** [comment]
That seems odd. Why would a bound function be different?

**Pull request closed without merging**